### ISSUE TITLE
dsc-pileup run extremely slowly for some samples?

### ISSUE THREAD

**Issue by allumette** [statgen/popscle#16]
Hi,
Recently I run freemuxlet on about 30 smart-seq samples with the following cmd, of which 2 samples run extremely slowly (our cluster can not run jobs longer than 14 days). Can you give me some suggestions?

popscle dsc-pileup --sam input.bam --vcf 1000G_phase1.snps.high_confidence.hg38.exons.vcf --out out_dir/pileup --tag-group RG --tag-UMI UB
popscle freemuxlet --plp out_dir/pileup --out out_dir/snv --nsample 4

input.bam is a smart-seq bam file generated by STAR, so its tag-group is RG, and there is no UB. I found freemuxlet perform very well with --tag-UMI UB set for smart-seq samples.

**Comment by Griffan** [comment]
Could you also post part of the log figure content so that we can locate at which step it took long time?

**Comment by allumette** [comment]
Patient1_BM_1M_PBSCT_1.pass.log is a sample which run relatively fast,
Normal_B_cell_12.log is still running (the slowest sample). As the data is read, it runs slower and slower.

**Comment by allumette** [comment]
Patient1_BM_1M_PBSCT_1.pass.log is a sample which run relatively fast,
Normal_B_cell_12.log is still running (the slowest sample). As the data is read, it runs slower and slower.

**Comment by allumette** [comment]
Patient1_BM_1M_PBSCT_1.pass.log is a sample which run relatively fast,
Normal_B_cell_12.log is still running (the slowest sample). As the data is read, it runs slower and slower.

**Comment by allumette** [comment]
[Normal_B_cell_12.log](https://github.com/statgen/popscle/files/3782709/Normal_B_cell_12.log) is still running (the slowest sample). As the data is read, it runs slower and slower.
[Patient1_BM_1M_PBSCT_1.pass.log](https://github.com/statgen/popscle/files/3782710/Patient1_BM_1M_PBSCT_1.pass.log) is a sample which run relatively fast,

**Comment by hyunminkang** [comment]
It might be related to the memory. Current version of dsc-pileup can use a
large amount of memory.

Hyun.
-----------------------------------------------------
Hyun Min Kang, Ph.D.
Associate Professor of Biostatistics
University of Michigan, Ann Arbor
Email : hmkang@umich.edu

**Comment by allumette** [comment]
Yes, I found it. The memory used is directly related to the number of barcodes and snvs. The two samples I run contain 92 barcodes and reference vcf file contain about 970k variants, which will not cost more than 24GB memory. I try to run them with 24GB, 48GB and 256GB memory respectively and the time consumed will not change. So there may be other reasons I think.

**Comment by hyunminkang** [comment]
If there is no UMI, it will consider every read as unique (which may not be
ideal). Are you sure that RG is cell barcode? What are the number of reads
and cells and how long does it take?

Thanks,
Hyun.
-----------------------------------------------------
Hyun Min Kang, Ph.D.
Associate Professor of Biostatistics
University of Michigan, Ann Arbor
Email : hmkang@umich.edu

**Comment by allumette** [comment]
1. Our samples are smart-seq, they do not have UMI.
2. We mapped reads  to get each bam file for a cell, named as 'barcode'.bam, then merged them to a single bam for a sample by samtools. So samtools wrote the filename (i.e. barcode) to RG.
3.  Patient1_BM_1M_PBSCT_1 has 59579770 reads and 92 cells in total, took 20.5 hours.
Normal_B_cell_12 has 113719054 reads and 92 cells in total, has taken 10 days and is still running.

**Comment by hyunminkang** [comment]
Hmm.. This doesn't seem right. Can you show how the BAM file looks like?

Hyun.
-----------------------------------------------------
Hyun Min Kang, Ph.D.
Associate Professor of Biostatistics
University of Michigan, Ann Arbor
Email : hmkang@umich.edu

**Comment by allumette** [comment]
[example.sam.zip](https://github.com/statgen/popscle/files/3802532/example.sam.zip) contains the first 1000 and last 1000 lines of sam file of sample Normal_B_cell_12 and Patient1_BM_1M_PBSCT_1.

**Comment by Griffan** [comment]
I tried to run with your files, the program finished normally without any memory leak errors. When you run popscle with different memory settings, have you recorded the actually usage of the memory during each run?

**Comment by allumette** [comment]
[c01b02n01_lilsh_Tue_Nov_12_10_03_12_CST_2019.pdf](https://github.com/statgen/popscle/files/3834129/c01b02n01_lilsh_Tue_Nov_12_10_03_12_CST_2019.pdf) is a record of first 20 min of dsc-pileup running on sample Patient1_BM_1M_PBSCT_1. 0.1%MEM is 250MB.
I have no full record of the two samples, but I gave each ot them 48GB mem to use. I have set mem limit for slurm system. If a task run out of memory within a linux swap size, it will print a warning, else it will be killed. I did not see warnings for the two samples.

**Comment by ghuls** [comment]
@allumette Check https://github.com/statgen/popscle/issues/22, it might help for your case too. I designed it for speeding up `dsc-pileup` but it might have a much lower memory usage too.
